pkg/operator/watchdog/cmd.go

```diff
@@ -0,0 +1,247 @@
+	defer shutdown()
+	defer close(pidObservedCh)
+	currentPID := 0
+	for {
+		// get the monitored process PID
+		observedPID, found, err := o.getPIDByNameFn(o.ProcessName)
+		if !found {
+			klog.Warningf("Process %q does not have PID", o.ProcessName)
+			return
+		}
+		if err != nil {
+			klog.Warning(err)
+			return
+		}
+
+		if currentPID == 0 {
+			currentPID = observedPID
+			pidObservedCh <- observedPID
+		}
+
+		// when the PID change, signal shutdown
+		if currentPID != observedPID {
+			return
+		}
+
+		time.Sleep(1 * time.Second)
+	}
+}
+
+// readInitialFileContent reads the content of files specified.
+// This is needed by file observer.
+func (o *FileWatcherOptions) readInitialFileContent() error {
+	o.initialContent = map[string][]byte{}
+	for _, name := range o.Files {
+		content, err := ioutil.ReadFile(name)
+		if err != nil {
+			return err
+		}
+		o.initialContent[name] = content
+	}
+	return nil
+}
+
+// setFilesPathPrefix mutates the file list and prefix every file with /proc/PID/root.
+func setFilesPathPrefix(oldFiles []string, pid int) []string {
+	files := []string{}
+	for _, file := range oldFiles {
+		files = append(files, filepath.Join("/proc", fmt.Sprintf("%d", pid), "root", file))
+	}
+	return files
+}
+
+func (o *FileWatcherOptions) Run(ctx context.Context) error {
+	runCtx, shutdown := context.WithCancel(ctx)
+
+	pidObservedCh := make(chan int)
+	go o.startPidObserver(shutdown, pidObservedCh)
+
+	// Wait while we get the initial PID for the process
+	klog.Infof("Waiting for process %q PID ...", o.ProcessName)
+	currentPID := <-pidObservedCh
+
+	// Mutate path for specified files as '/proc/PID/root/<original path>'
+	// This means side-car container don't have to duplicate the mounts from main container.
+	// This require shared PID namespace feature.
+	o.Files = o.setFilesPathPrefixFn(o.Files, currentPID)
+
+	// Read initial file content. If shared PID namespace does not work, this will error.
+	if err := o.readInitialFileContent(); err != nil {
+		o.recorder.Warningf("FileChangeWatchdogFailed", "Reading initial file content failed: %v", err)
+		return fmt.Errorf("unable to read initial file content: %v", err)
+	}
+
+	o.recorder.Eventf("FileChangeWatchdogStarted", "Started watching files for process %s[%d]", o.ProcessName, currentPID)
+
+	observer, err := fileobserver.NewObserver(o.Interval)
+	if err != nil {
+		o.recorder.Warningf("ObserverFailed", "Failed to start to file observer: %v", err)
+		klog.Fatalf("unable to start file observer: %v", err)
+	}
+
+	observer.AddReactor(func(file string, action fileobserver.ActionType) error {
+		o.recorder.Eventf("FileChangeObserved", "Observed change in file %q, sending SIGINT to %s[%d]", file, o.ProcessName, currentPID)
+		// When file change is observed, don't give up on errors when we fail to send signal, but retry sending
+		// and fail hard if we fail to deliver it.
+		retries := 0
+		return wait.PollImmediate(1*time.Second, 5*time.Second, func() (done bool, err error) {
```

Comment:
what are these intervals? Why those numbers?

Reply:
I will add godoc (make 5 attempts to SIGINT, wait 1s between them... This might not be needed, but i'm skeptical that things will always work and this is a critical path for watchdog, hence the retry.